### PR TITLE
[NativeAOT] Remove WaitForSuspend and some ordering cleanups.

### PR DESCRIPTION
**Pull request by VSadov**
- Noone, except the suspending thread, needs to wait for suspension to complete. Other threads just need to leave coop mode, otherwise can run. The original purpose for `WaitForSuspend ` is not very clear. 
- More explicit ordering when working with `m_pTransitionFrame`. 
We only need to enforce compiler ordering, but it is important.
- Process wide barriers around GC for weak architectures (since preemptive threads do not synchronize with GC).

### CODE REVIEW

**Comment by jkotas** [comment]
> The original purpose for WaitForSuspend is not very clear.

It was to ensure that we never call `SuspendThread` on a thread that is not running code under our control. In CoreCLR suspend thread implementation, it is possible for us to call SuspendThread on a thread that has been executing deep inside a PInvoke. It is  potential reliability problem. If we call `SuspendThread` on a thread that is executing in certain places in Windows, it can result in deadlocks. The probability of this happening is very low, but it can happen. Redhawk was designed to higher reliability bar than classic CLR and so it had handling for this theoretical problem.

**Comment by VSadov** [comment]
> It was to ensure that we never call SuspendThread on a thread that is not running code under our control.

I do not think that is achievable, considering various races. It happens all the time and that is why the very first thing we do after interrupting a thread is checking if it runs managed code.  

The idea behind `WaitForSuspend` might have worked with some additional constraints I am not aware, but what I see right now - a thread that transitions to preemptive, after posting a frame, checks whether a suspension is in progress, and if it is unlucky to see suspension in progress, then it would wait until suspension is over and GC starts, and then it continue with its unmanaged business. 

The wait is accidental and provides no value. Waiting on an event is still unmanaged code, the thread could just continue running its thing. We would have noticed when we tried to interrupt it, and let it go just the same.

**Comment by jkotas** [comment]
> Waiting on an event is still unmanaged code, the thread could just continue running its thing.

I agree with you that waiting on an event is still unmanaged code. I guess that the wait on an event code path was reviewed to not have any of those potential issues with Suspend/Resume being called on it. This change exposes all code that may be running inside PInvoke to the potential issues with Suspend/Resume being called on it.

File: src/coreclr/nativeaot/Runtime/threadstore.cpp [hunk 1]
```diff
@@ -272,6 +276,16 @@ void ThreadStore::ResumeAllThreads(bool waitForGCEvent)
     }
     END_FOREACH_THREAD
 
+#if defined(TARGET_ARM) || defined(TARGET_ARM64)
+        // Flush the store buffers on all CPUs, to ensure that they all see changes made
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/threadstore.cpp:280]: Nit: Extra indentation

**Review by jkotas — APPROVED** [review]
I do believe that the current native AOT implementation is more robust that what we have in CoreCLR. However, we should not need to be building native AOT thread suspension to higher robustness bar, so I am ok with deleting the WaitForSuspend logic if you still believe that it is a good idea.

**Comment by VSadov** [comment]
Let me think about this a bit more. 

So far it seems that we do not take advantage of additional guarantees and just expect that threads may be interrupted anywhere and support for that should be sufficient (or so I believe).   
In such case simplifying PInvoke and being prepared to handle more cases in suspension seems like a good trade, because PInvoke is more frequent.

**Comment by VSadov** [comment]
Considering that we are expecting this to run on multiple platforms and architectures, the additional guarantee provided is not a lot. Even before this change, while debugging I have seen code interrupted in quite a lot of situations - in syscalls, in implementation of OS locks/mutexes, in signal dispatch, in exception handling, in code that cannot provide reliable CONTEXT (we check and retry in such cases).

Having one more lock on suspension path is also a cause of some concern - need to think about not get into some kind of reentrancy.

I think we should take this change.

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**